File: src/python/flashloan.py
import os
from web3 import Web3
from dotenv import load_dotenv

# Load environment variables
load_dotenv()

SEPOLIA_URL = os.getenv("Sepolia_Url")
WALLET_ADDRESS = os.getenv("Wallet")
PRIVATE_KEY = os.getenv("Private_Key")
SMART_CONTRACT_ADDRESS = os.getenv("Smart_Contract")
USDT_ADDRESS = os.getenv("Sepolia_USDT")
USDC_ADDRESS = os.getenv("Sepolia_USDC")
SMART_CONTRACT_ADDRESS_2 = os.getenv("Smart_Contract_2")

# Flash Loan Contract ABI
CONTRACT_ABI = [
    {
        "inputs": [
            {"internalType": "address", "name": "_token", "type": "address"},
            {"internalType": "uint256", "name": "_amount", "type": "uint256"}
        ],
        "name": "fn_RequestFlashLoan",
        "outputs": [],
        "stateMutability": "nonpayable",
        "type": "function"
    }
]

# Connect to Sepolia network
web3 = Web3(Web3.HTTPProvider(SEPOLIA_URL))

# Load smart contract
contract = web3.eth.contract(address=SMART_CONTRACT_ADDRESS, abi=CONTRACT_ABI)

def request_flash_loan(token_address, raw_amount):
    try:
        print(f"Requesting flash loan for {raw_amount} of token at {token_address}...")

        # Adjust amount for USDT/USDC (6 decimals)
        amount = raw_amount * 10**6
        
        # Get nonce for the transaction (to avoid duplicate transactions)
        nonce = web3.eth.get_transaction_count(WALLET_ADDRESS)

        # Build transaction for the flash loan request
        transaction = contract.functions.fn_RequestFlashLoan(token_address, amount).build_transaction({
            'chainId': 11155111,  # Sepolia chain ID
            'gas': 300000,        # Set a gas limit
            'gasPrice': web3.to_wei('10', 'gwei'),  # Gas price in Gwei
            'nonce': nonce,
            'from': WALLET_ADDRESS
        })

        # Sign the transaction
        signed_txn = web3.eth.account.sign_transaction(transaction, private_key=PRIVATE_KEY)

        # Send the signed transaction
        tx_hash = web3.eth.send_raw_transaction(signed_txn.rawTransaction)

        # Wait for the transaction receipt
        receipt = web3.eth.wait_for_transaction_receipt(tx_hash)

        print(f"Transaction successful with hash: {tx_hash.hex()}")
        return receipt

    except Exception as e:
        print(f"Error triggering flash loan: {e}")

# Example usage: Request a flash loan for 1000 USDT or USDC
if __name__ == "__main__":
    TOKEN_ADDRESS = USDC_ADDRESS  # Change to USDT_ADDRESS as needed
    RAW_AMOUNT = 2000  # Example raw amount, 1000 USDT/USDC
    request_flash_loan(TOKEN_ADDRESS, RAW_AMOUNT)

File: src/javascript/flashloan.js
require('dotenv').config(); // Load environment variables
const {Web3} = require('web3');

// Load environment variables from .env file
const SEPOLIA_URL = process.env.Sepolia_Url; // Sepolia network URL
const WALLET_ADDRESS = process.env.Wallet; // Wallet address
const PRIVATE_KEY = process.env.Private_Key; // Private key
const SMART_CONTRACT_ADDRESS = process.env.Smart_Contract; // Flash Loan Smart Contract address
const USDT_ADDRESS = process.env.Sepolia_USDT; // USDT contract address
const USDC_ADDRESS = process.env.Sepolia_USDC; // USDC contract address
const SMART_CONTRACT_ADDRESS_2 = process.env.Smart_Contract_2;

// Flash Loan Contract ABI
const CONTRACT_ABI = [
    {
        "inputs": [
            { "internalType": "address", "name": "_token", "type": "address" },
            { "internalType": "uint256", "name": "_amount", "type": "uint256" }
        ],
        "name": "fn_RequestFlashLoan",
        "outputs": [],
        "stateMutability": "nonpayable",
        "type": "function"
    }
];

// Connect to the Sepolia network
const web3 = new Web3(SEPOLIA_URL);

// Load the smart contract
const contract = new web3.eth.Contract(CONTRACT_ABI, SMART_CONTRACT_ADDRESS_2);

// Function to trigger the flash loan
async function requestFlashLoan(tokenAddress, rawAmount) {
    try {
        console.log(`Requesting flash loan for ${rawAmount} of token at ${tokenAddress}...`);

        // Since USDT and USDC have 6 decimals, multiply rawAmount by 10^6
        const amount = rawAmount * (10 ** 6);
        
        const nonce = await web3.eth.getTransactionCount(WALLET_ADDRESS, 'latest'); // Get the nonce

        // Build transaction
        const tx = {
            from: WALLET_ADDRESS,
            to: SMART_CONTRACT_ADDRESS_2,
            gas: 300000, // Adjust gas limit based on your contract's complexity
            gasPrice: web3.utils.toWei('10', 'gwei'), // Adjust gas price
            data: contract.methods.fn_RequestFlashLoan(tokenAddress, amount).encodeABI(),
            nonce: nonce
        };

        // Sign the transaction
        const signedTx = await web3.eth.accounts.signTransaction(tx, PRIVATE_KEY);

        // Send the transaction
        const receipt = await web3.eth.sendSignedTransaction(signedTx.rawTransaction);
        console.log(`Transaction successful! Hash: ${receipt.transactionHash}`);
    } catch (error) {
        console.error("Error during flash loan request:", error.message);
    }
}

// Example Usage: Request a flash loan for 1000 USDT or USDC
const TOKEN_ADDRESS = USDT_ADDRESS; // Change to `USDC_ADDRESS` or `USDT_ADDRESS` as needed
const RAW_AMOUNT = 2000; // The raw amount, e.g., 1000 USDT/USDC (before adjusting for decimals)

requestFlashLoan(TOKEN_ADDRESS, RAW_AMOUNT);
